docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,409 @@
+
+So this needs to be cached and refreshed whenever a new operator is installed.
+
+
+### Fetch ClusterServiceVersion to generate the information
+
+If none of the above approaches work, we finally fallback to getting the information from `ClusterServiceVersion`
+
+We generate the description in this approach based on `spec.customresourcedefinitions` present in the ClusterServiceVersion.
+This CRD def is different from what is provided by the `CustomResourceDefinition` as it has less information.
+
+This is how one of the `customresourcedefinition` looks like ( Kafka from strimzi )
+```
+{
+  "parameters": [
+    {
+      "description": "Kafka version",
+      "displayName": "Version",
+      "path": "kafka.version",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:text"
+      ]
+    },
+    {
+      "description": "The desired number of Kafka brokers.",
+      "displayName": "Kafka Brokers",
+      "path": "kafka.replicas",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:podCount"
+      ]
+    },
+    {
+      "description": "The type of storage used by Kafka brokers",
+      "displayName": "Kafka storage",
+      "path": "kafka.storage.type",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:select:ephemeral",
+        "urn:alm:descriptor:com.tectonic.ui:select:persistent-claim",
+        "urn:alm:descriptor:com.tectonic.ui:select:jbod"
+      ]
+    },
+    {
+      "description": "Limits describes the minimum/maximum amount of compute resources required/allowed",
+      "displayName": "Kafka Resource Requirements",
+      "path": "kafka.resources",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:resourceRequirements"
+      ]
+    },
+    {
+      "description": "The desired number of Zookeeper nodes.",
+      "displayName": "Zookeeper Nodes",
+      "path": "zookeeper.replicas",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:podCount"
+      ]
+    },
+    {
+      "description": "The type of storage used by Zookeeper nodes",
+      "displayName": "Zookeeper storage",
+      "path": "zookeeper.storage.type",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:select:ephemeral",
+        "urn:alm:descriptor:com.tectonic.ui:select:persistent-claim",
+        "urn:alm:descriptor:com.tectonic.ui:select:jbod"
+      ]
+    },
+    {
+      "description": "Limits describes the minimum/maximum amount of compute resources required/allowed",
+      "displayName": "Zookeeper Resource Requirements",
+      "path": "zookeeper.resources",
+      "x-descriptors": [
+        "urn:alm:descriptor:com.tectonic.ui:resourceRequirements"
+      ]
+    }
+  ]
+}
+```
+
+## Allowing user to set the parameters 
+
+The current approach of sending map values via cli are as follows - 
+
+- we would add a `-p` cobra param as a list
+- each of the parameter would represent the key in a map and value in a map `e.g. -p "key"="value"`
```

Comment:
`key` needs to match "parameter name" as shown in `odo catalog component describe`

Reply:
adding this

Reply:
actually if you think about it - 
the openapi json output would look like this
https://gist.github.com/girishramnani/0d56b28108594657a25a002993840872
So there would be nested structures. 
on the other hand `x-descriptor` descriptions are flat as they use `json-path` like syntax to specify values.

From readability standpoint `x-descriptors` is better as a user can just get the `key` from the description and use it in `odo service create`

but the json output of `x-descriptor` is limiting.

Reply:
@kadel I dont know if we can flatten this `json-schema` structure to something human readable.
there are some arrays in the schema as well so we cannot show arrays as a `json-schema`
e.g.

we have a `secret` list of objects defined in the json schema. if we try to flatten it - we cannot have it like this `zookeeper.secret[0].name` its not always gonna be`0` indexed? or we just make it `zookeeper.secret.name` and assume that its append only list? or `zookeeper.secret[-1].name` like how sjson prefer?

Reply:
> @kadel I dont know if we can flatten this `json-schema` structure to something human readable.

you mean for example this form `zookeeper.secret.name`? I think that this is perfectly human readable



> we have a `secret` list of objects defined in the json schema. if we try to flatten it - we cannot have it like this `zookeeper.secret[0].name` its not always gonna be`0` indexed?

Why it wouldn't be 0 indexed?

> or we just make it zookeeper.secret.name and assume that its append only list?

I'm afraid that this wouldn't work. 
Ther will be problems if you have array of objects. Like for example `env`


`-p spec.env.name FOO -p spec.env.value BAR`

Would this resolve to 
```
spec:
  env:
    - name: FOO
      value: BAR
```

or 


```
spec:
  env:
    - name: FOO
    - value: BAR
```

?

Reply:
I think that we will have to use indexes like `zookeeper.secret[0].name`

`-p spec.env[0].name FOO -p spec.env[0].value BAR`

```
spec:
  env:
    - name: FOO
      value: BAR
```

`-p spec.env[0].name FOO -p spec.env[1].value BAR`
```
spec:
  env:
    - name: FOO
    - value: BAR
 ```

But we need to figure out how to communicate to user that the given field is array and the index is needed in `odo catalog service describe` output

Reply:
Yeah as I think about it, append only won’t work.

and the reason I said. [0] won’t work I meant that the idea that key from describe cannot be just copied as is to odo service create.
The user has to think that this is an array so it would have a [index].
Maybe we can show it as `secret[*].name` and explain in describe that * needs to be an index 0,1….

Reply:
> `-p spec.env[0].name FOO -p spec.env[0].value BAR`

Do you folks think this would be good from UX point of view? Taking a step back, why can't we rely only on swagger.json for params backed services? If the user doesn't want to pass parameters, we can use what's there in alm-examples.

Correct me if I am misinterpreting things here.

Reply:
That is the case, if the user doesn’t provide any params we use alm-examples

Reply:
But if the user does want to provide params, they will have to use the format like `-p spec.env[0].name FOO -p spec.env[0].value BAR`? This is a horrible UX, IMO.

Reply:
This is only for lists and I am open to suggestions on the format

Reply:
It would be really amazing if we can confirm this approach 🙂

Reply:
I can't currently think of any other solutions or this. @dharmit feel free to propose an alternative.

Reply:
An alternative could be to not use the array formatting and check in odo code if the parameter that the user's trying to set is indeed available. odo would have to do some heavy lifting in that case in terms of validating stuff.

But this is implementation detail, and it's not relevant for proposal.